addons/mass_mailing/controllers/main.py

```diff
@@ -162,3 +162,51 @@ def send_feedback(self, mailing_id, res_id, email, feedback, token):
                 record.sudo().message_post(body=_("Feedback from %s: %s" % (email, feedback)))
             return bool(records)
         return 'error'
+
+    @http.route(['/mail/mailing/<int:mailing_id>/browser_view'], type='http', website=True, auth='public')
+    def browser_view(self, mailing_id, email=None, res_id=None, token=""):
+        mailing = request.env['mailing.mailing'].sudo().browse(mailing_id)
+        if mailing.exists():
+
+            res_id = res_id and int(res_id)
+            if not self._valid_unsubscribe_token(mailing_id, res_id, email, str(token)) and not request.env.user.has_group('mass_mailing.group_mass_mailing_user'):
+                raise exceptions.AccessDenied()
+
+            res = mailing.convert_links()
+
+            base_url = request.env['ir.config_parameter'].sudo().get_param('web.base.url')
+            unsubscribe_new_url = werkzeug.urls.url_join(
```

Comment:
What about moving _get_unsubscribe_url from mail_mail to mailing.mailing model, and calling it here with res_id, email parameters ? Same for _get_browser_view_url that could be relocated on mailing and used here and in mail_mail as we have mailing_id on mail_mail record.

This could would be better located on mailing model, what do you think ?